Changed flycheck-rust-manifest-directory to use 'cargo metadata'

kngwyu opened the pull request:
This commit fixes https://github.com/flycheck/flycheck/issues/1397.
By the recent change of cargo(https://github.com/rust-lang/cargo/pull/4788), flycheck can't detect error files in multi-crate project of Rust.
So I changed to use `workspace_root` in `cargo metadata` 's outputs, as working directory.
If the user's cargo is a bit old and there's no 'workspace_root', `flycheck-rust-manifest-directory` search a directory with `Cargo.toml`, as ever.

I've seen @fmdkdd's comment in https://github.com/rust-lang/cargo/issues/4998, but I wrote this patch just because I need this, so please feel free to abort this PR if you don't like.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/flycheck/flycheck?pullRequest=1422) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/flycheck/flycheck?pullRequest=1422) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/flycheck/flycheck?pullRequest=1422) it.</sub>

fmdkdd commented:
Thanks for reminding me about this.  It looks like we don't have a choice, we have to call `cargo metadata` after all, as the alternative rust-lang/rust#47939 has been dropped.

And thank you for writing the patch!  I'll try to review it soon.

kngwyu commented:
I'm sorry I checked this change cause 
```
Suspicious state from syntax checker rust-cargo: Flycheck checker rust-cargo returned non-zero exit code 101, but its output contained no errors: error: no library targets found
```
in the project like
```
├── Cargo.toml
├── crate1
│   ├── Cargo.lock
│   ├── Cargo.toml
│   ├── src
│   └── target
├── crate2
│   ├── Cargo.toml
│   └── src
├── rustfmt.toml
├── target
```
so it needs to be fixed!

kngwyu commented:
This error occurs in a Project which has both `bin` and `lib` (and etc..).
In addition, changing `working-directory` into `workspace_root` can make compile time much more longer!
Hmm, now I don't feel this is a good way....

Is there any good idea to separate the directory where flycheck run command, from working directory?

fmdkdd commented:
> Is there any good idea to separate the directory where flycheck run command, from working directory?

The working directory *is* the directory where flycheck runs the command.

But for fixing #1397, the only thing that we need to change is to expand the names relatively to the workspace root, whereas previously they were relative to the working directory.  We don't want to change the working directory.

So the change should probably rather go to `:error-filter` or `:error-parser`.

kngwyu commented:
@fmdkdd 
Thanks for your advice.
I'll try it.

kngwyu commented:
I modified to use completely new way.
It works well in my test directory, but I'm not sure it's a good way.

fmdkdd commented:
@kngwyu Thanks!  That's closer to what the fix should be.  I agree that we should expand into an absolute path, only if the workspace root is present.

But doing the expansion in `flycheck-parse-rustc-diagnostic` will break the plain `rust` checker, which uses the same parser.  I think we should rather do the expansion into absolute paths in the  `:error-filter` of `rust-cargo`, so we don't mess with parsing and the other checker.

I've pushed some changes in that direction on your branch.  I've also fixed the tests, and added a new test for the workspace situation.  It works with nightly and stable.

kngwyu commented:
@fmdkdd Oh, I didn't consider `rust` checker, sorry, and thanks for your all work!

fmdkdd commented:
@kngwyu No worries, that's what reviews are for.

I've changed `alist-get` to `let-alist` as well, since `alist-get` is not available in Emacs 24.  I think we should be good.

@cpitclaudel Can you take a look?

cpitclaudel commented on flycheck.el:
```diff
@@ -9480,6 +9480,26 @@ or if the current buffer has no file name."
        (locate-dominating-file buffer-file-name "Cargo.toml")))
 
+(defun flycheck-rust-cargo-metadata ()
+  "Exec 'cargo metadata' and return the result as parsed JSON object.
```
I'd say `Run` rather than `Exec`, but that's nitpicking :)

cpitclaudel commented on flycheck.el:
```diff
@@ -9480,6 +9480,26 @@ or if the current buffer has no file name."
+(defun flycheck-rust-cargo-metadata ()
+  "Exec 'cargo metadata' and return the result as parsed JSON object.
+
+Return nil if the command failed or if a parse error occured."
```
Do we really want to fail silently for failed calls to `cargo metadata` and for parse errors? (IIUC this is for backwards compatibility, but I'm afraid it'll create harder-to-debug bugs if things change unexpectedly)

kngwyu replied:
I think it's only because 'cargo' 's bug if parse error occurs, so we can allow error here, but I'm not sure.

cpitclaudel commented on flycheck.el:
```diff
@@ -9480,6 +9480,26 @@ or if the current buffer has no file name."
+  "Exec 'cargo metadata' and return the result as parsed JSON object.
+
+Return nil if the command failed or if a parse error occured."
+  (let ((metadata-lines
```
`-when-let` here would remove the need for the `and` below

cpitclaudel commented on flycheck.el:
```diff
@@ -9480,6 +9480,26 @@ or if the current buffer has no file name."
+           (process-lines "cargo" "metadata" "--no-deps" "--format-version" "1"))))
+    (and metadata-lines
+         (ignore-errors
+           (json-read-from-string (string-join metadata-lines ""))))))
```
That call to string-join should use `\n`, not ``, I think (though it might not matter in this particular case?)
Besides, would it read nicer to use `call-process` and `json-read` instead of `process-lines` + `string-join` + `json-read-from-string`?

fmdkdd replied:
In this case, we may just use `flycheck-parse-json`?

kngwyu replied:
@fmdkdd 
`flycheck-parse-json` returns a list of parsed JSON lines, so isn't it too complex for this case, where what we want is only `metadata` ?

fmdkdd replied:
It may be overkill, yes.  Using `call-process` and `json-read` directly is fine :)

cpitclaudel commented on flycheck.el:
```diff
@@ -9480,6 +9480,26 @@ or if the current buffer has no file name."
+Return nil if the workspace root does not exist (for Rust
+versions inferior to 1.25)."
+  (let ((metadata (flycheck-rust-cargo-metadata)))
+    (let-alist metadata
```
No need to bind `metadata` here, I think (`(let-alist (flycheck-rust-cargo-metadata) …)`)

kngwyu commented:
Thanks for review.
I modified a bit, but for `flycheck-rust-cargo-metadata` I'm not sure how we should handle errors and parse json.

kngwyu commented:
I simplified `flycheck-rust-cargo-metadata` and changed not to use `ignore-errors` in this function.
The newest option `--no-deps` in  `cargo metadata --no-deps --format-version 1` was added [2 years ago](https://github.com/rust-lang/cargo/commit/80e3c05beefad3a19b2cd694d0acfd4e09041fb0#diff-24a11135fb787144156765e41edea25b), so backwards compatibility is OK.
Thus function never fail unless `cargo` doesn't exist or breaking change is done.

fmdkdd commented on flycheck.el:
```diff
@@ -9480,6 +9480,24 @@ or if the current buffer has no file name."
+(defun flycheck-rust-cargo-metadata ()
+  "Run 'cargo metadata' and return the result as parsed JSON object.
+
+Return nil if the command failed or if a parse error occured."
```
I guess this line can go now that we don't ignore errors anymore.

fmdkdd replied:
Oh and it looks like the file is misformatted.  Please run `make format`, which should take care of the issue.

kngwyu commented:
Modified.
Thanks for your review!

fmdkdd commented on flycheck.el:
```diff
@@ -9483,7 +9483,7 @@ or if the current buffer has no file name."
   "Run 'cargo metadata' and return the result as parsed JSON object.
 
-Return nil if the command failed or if a parse error occured."
+We don't ignore any errors in this function."
```
I don't think we need to spell out that we don't do any error catching, as that is a detail of implementation.  Anyone caring about implementation can see in the code that no error checking is done.

Please remove this line :)

kngwyu commented:
Modified.

fmdkdd commented:
Great!  Thank you very much for this patch! :clap:
